### PR TITLE
Add global task queue support for distributed workers

### PR DESCRIPTION
**Pull request by panghy**
**Summary:**
Implements centralized (global) task queue support enabling background workers (build and maintenance) to run in a separate process and service multiple vector index instances from a single shared pool. When enabled, local task queues and workers are bypassed, allowing for true distributed processing.

**Changes:**
- **Proto definitions**: Added `GlobalBuildTask` and `GlobalMaintenanceTask` wrapper messages that carry index directory paths alongside original task payloads
- **Configuration**: Introduced `GlobalTaskQueueConfig` to hold shared build and maintenance queues; added optional `globalTaskQueueConfig` to `VectorIndexConfig`
- **Adapters**: Created `GlobalBuildQueueAdapter` and `GlobalMaintenanceQueueAdapter` that wrap per-index tasks with directory paths and delegate to global queues
- **Index integration**: Modified `FdbVectorIndex.createOrOpen()` to use adapters when global mode is enabled; local workers are skipped
- **Worker runner**: Implemented `GlobalWorkerRunner` — a standalone programmatic API that claims tasks from global queues, resolves index directories, reconstructs per-index configs from persisted `IndexMeta`, and dispatches to `SegmentBuildService` and `MaintenanceService`
- **Serializers**: Added `GlobalBuildTaskSerializer` and `GlobalMaintenanceTaskSerializer` for proto serialization
- **Tests**: Comprehensive integration tests covering multi-index scenarios, FIFO processing, and edge cases; all tests pass with coverage thresholds met

**Technical Details:**
- Data-format parameters (dimension, metric, pqM, pqK, graphDegree, etc.) are read from each index's persisted `IndexMeta`
- Operational parameters (compaction knobs, vacuum thresholds, transaction limits) come from a uniform template config
- Backward compatible — existing behavior unchanged when `globalTaskQueueConfig` is not set
- Uses sentinel tasks (seg_id < 0) for graceful worker shutdown

**Testing:**
- `GlobalTaskQueueIntegrationTest`: Multi-index enqueue routing, build/maintenance processing, FIFO ordering
- `GlobalTaskQueueEdgeCaseTest`: Serializer round-trips, concurrent operations, error handling
- All existing tests continue to pass; coverage thresholds maintained (≥90% line, ≥75% branch)

**Checklist:**
- [x] Proto messages and serializers implemented
- [x] `VectorIndexConfig` and `GlobalTaskQueueConfig` integration
- [x] Adapter queues route tasks correctly
- [x] `GlobalWorkerRunner` processes tasks from multiple indices
- [x] Integration tests cover all acceptance criteria
- [x] Build passes: `./gradlew build` with coverage verification
- [x] Code formatted: `./gradlew spotlessApply`

---
Pull Request opened by [Augment Code](https://www.augmentcode.com/) with guidance from the PR author

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -155,18 +162,81 @@ Background compactions are planned from small SEALED segments and throttled by `
 - Programmatic vacuum:
   - Use `MaintenanceService.vacuumSegment(segId, minDeletedRatio)` to run a targeted vacuum now.
   - Or enqueue and process via the maintenance queue using `MaintenanceWorker`.
-  - Example:
-    ```java
-    var dirs = FdbDirectories.openIndex(root, db).get(5, TimeUnit.SECONDS);
-    var svc = new MaintenanceService(cfg, dirs);
-    svc.vacuumSegment(0 /* segId */, 0.0 /* minDeletedRatio */).get(10, TimeUnit.SECONDS);
-    ```
+  - Example:`var dirs = FdbDirectories.openIndex(root, db).get(5, TimeUnit.SECONDS); var svc = new MaintenanceService(cfg, dirs); svc.vacuumSegment(0 /* segId */, 0.0 /* minDeletedRatio */).get(10, TimeUnit.SECONDS);`
```
[the remark this answers is not in the document]
Comment (panghy) [reply README.md]: Fixed in 6844933 — restored the programmatic vacuum example as a fenced Java code block with proper spacing after `Example:`.

File: README.md [hunk 2]
```diff
@@ -155,18 +162,81 @@ Background compactions are planned from small SEALED segments and throttled by `
...
+// Template config for operational settings (data params read from each index's persisted IndexMeta)
+VectorIndexConfig templateConfig = VectorIndexConfig.builder(db, dummyDir)
+    .dimension(1) // overridden per-index from IndexMeta
+    .buildTxnLimitBytes(5_000_000)
+    .maxConcurrentCompactions(2)
+    .build();
+
+// Start workers: 4 build threads + 2 maintenance threads
+GlobalWorkerRunner runner = new GlobalWorkerRunner(db, globalCfg, templateConfig, 4, 2);
+runner.start();
```
[the remark this answers is not in the document]
Comment (panghy) [reply README.md]: Verified — the README snippet already matches the current public API: `new GlobalWorkerRunner(db, workerConfig, globalCfg)` constructor and `runner.start(4, 2)` startup. No changes needed here.

File: src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java [hunk 3]
```diff
@@ -91,62 +69,82 @@ private VectorIndexConfig(Builder b) {
...
-      throw new IllegalArgumentException("buildTxnSoftLimitRatio must be in (0,1)");
-    if (b.buildSizeCheckEvery <= 0) throw new IllegalArgumentException("buildSizeCheckEvery must be positive");
-    this.buildTxnLimitBytes = b.buildTxnLimitBytes;
-    this.buildTxnSoftLimitRatio = b.buildTxnSoftLimitRatio;
-    this.buildSizeCheckEvery = b.buildSizeCheckEvery;
+
+    // Build or use the provided WorkerConfig for all operational settings
+    if (b.workerConfig != null) {
+      this.workerConfig = b.workerConfig;
+    } else {
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java:81]: Fixed in 607e2fe — `WorkerConfig`'s constructor now validates all operational invariants (same constraints as `VectorIndexConfig`), so a pre-built `WorkerConfig` cannot bypass validation. Both construction paths now enforce identical constraints.

File: src/main/java/io/github/panghy/vectorsearch/config/WorkerConfig.java [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+      return this;
+    }
+
+    public Builder defaultGraphAlpha(double v) {
+      this.defaultGraphAlpha = v;
+      return this;
+    }
+
+    /** Builds the immutable {@link WorkerConfig}. */
+    public WorkerConfig build() {
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/config/WorkerConfig.java:451]: Fixed in 607e2fe — Added full invariant validation in `WorkerConfig(Builder)` matching all constraints historically enforced by `VectorIndexConfig`: non-negative durations, ratio ranges `[0,1]` and `(0,1)`, min/max relationships (e.g., `compactionMaxSegments >= compactionMinSegments`), positive counts, and null checks. Added comprehensive tests covering each invalid case.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    });
+  }
+
+  private CompletableFuture<Void> processMaintenanceTask(
+      MaintenanceTask t, VectorIndexConfig cfg, IndexDirectories dirs) {
+    MaintenanceService svc = new MaintenanceService(cfg, dirs);
+    if (t.hasVacuum()) {
+      var v = t.getVacuum();
+      return svc.vacuumSegment(v.getSegId(), v.getMinDeletedRatio());
+    }
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java:233]: Addressed on the current head (`9866a41`). In global-queue mode, follow-up maintenance work from vacuum processing is now routed through the global maintenance queue rather than a per-index local queue, so `FindCompactionCandidates` work remains visible to `GlobalWorkerRunner`.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        List.copyOf(indexPath),
+        path -> DirectoryLayer.getDefault()
+            .open(db, path)
+            .thenCompose(indexDir -> FdbDirectories.openIndex(indexDir, db)));
+  }
+
+  /**
+   * Reads persisted {@link IndexMeta} from the index and reconstructs a {@link VectorIndexConfig}
+   * using data params from the meta and operational settings from the template.
+   */
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java:363]: Fixed on the current head (`9866a41`). `buildConfigForIndex(...)` now carries over the full operational `WorkerConfig` settings needed for maintenance behavior, and the docs/comments were tightened so the reconstructed config description matches what is actually applied.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    if (t.hasVacuum()) {
+      var v = t.getVacuum();
+      return svc.vacuumSegment(v.getSegId(), v.getMinDeletedRatio());
+    }
+    if (t.hasCompact()) {
+      return svc.compactSegments(t.getCompact().getSegIdsList());
+    }
+    if (t.hasFindCandidates()) {
+      return svc.findCompactionCandidates(t.getFindCandidates().getAnchorSegId())
+          .thenApply(v -> null);
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java]: Fixed on the current head (`9866a41`). Global `FindCompactionCandidates` handling now mirrors the local maintenance flow: it applies throttling/in-flight checks, marks chosen segments `COMPACTING`, and enqueues the follow-up `Compact` task. Added focused edge-case coverage for this path as well.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+          .thenApply(v -> true);
+    });
+  }
+
+  private CompletableFuture<Void> processMaintenanceTask(
+      MaintenanceTask t, VectorIndexConfig cfg, IndexDirectories dirs) {
+    MaintenanceService svc = new MaintenanceService(cfg, dirs);
+    if (t.hasVacuum()) {
+      var v = t.getVacuum();
+      return svc.vacuumSegment(v.getSegId(), v.getMinDeletedRatio());
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java:232]: This is addressed on the current head (`9866a41`). `MaintenanceService` now respects global-queue mode and routes follow-up maintenance work back into the global maintenance queue instead of creating/using a per-index local maintenance queue.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+  CompletableFuture<Boolean> runOnceBuild() {
+    return buildQueue.awaitAndClaimTask().thenCompose(claim -> {
+      GlobalBuildTask gbt = claim.task();
+      BuildTask bt = gbt.getTask();
+      // Sentinel: seg_id < 0 means shutdown signal
+      if (bt.getSegId() < 0) {
+        return claim.complete().thenApply(v -> true);
+      }
+      List<String> indexPath = gbt.getIndexPathList();
+      return resolveIndexDirs(indexPath)
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java:180]: Fixed on the current head (`9866a41`). `GlobalWorkerRunner` now validates wrapper payloads/index paths before resolving directories, so malformed or empty `GlobalBuildTask` / `GlobalMaintenanceTask` claims fail clearly instead of attempting to open an empty path. Added targeted edge-case tests for malformed wrappers.

**Comment by panghy** [comment]
All currently raised review items have been addressed on the latest head (`9866a41`). CI is running on that commit now; once it clears, this should be ready for re-review.

File: src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java [hunk 10]
```diff
@@ -91,62 +69,82 @@ private VectorIndexConfig(Builder b) {
...
+      // Validate operational fields before building WorkerConfig
+      if (b.estimatedWorkerCount <= 0)
+        throw new IllegalArgumentException("estimatedWorkerCount must be positive");
+      if (b.maxConcurrentCompactions < 0)
+        throw new IllegalArgumentException("maxConcurrentCompactions must be >= 0");
+      if (b.vacuumCooldown.isNegative()) throw new IllegalArgumentException("vacuumCooldown must be >= 0");
+      if (!(b.vacuumMinDeletedRatio >= 0.0 && b.vacuumMinDeletedRatio <= 1.0))
+        throw new IllegalArgumentException("vacuumMinDeletedRatio must be in [0,1]");
+      requirePositive(b.defaultTtl, "defaultTtl");
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java:91]: Fixed in `404c1ba`. Added `Objects.requireNonNull(b.vacuumCooldown, "vacuumCooldown must not be null")` and `Objects.requireNonNull(b.defaultThrottle, "defaultThrottle must not be null")` before the `isNegative()` checks in `VectorIndexConfig`'s constructor, consistent with `WorkerConfig`'s validation. Added focused tests for both null cases.

File: src/main/java/io/github/panghy/vectorsearch/config/WorkerConfig.java [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+    this.prefetchCodebooksEnabled = b.prefetchCodebooksEnabled;
+    this.prefetchCodebooksSync = b.prefetchCodebooksSync;
+    this.instantSource = b.instantSource;
+    this.metricAttributes = Map.copyOf(b.metricAttributes);
+
+    this.defaultMaxSegmentSize = b.defaultMaxSegmentSize;
+    this.defaultPqM = b.defaultPqM;
+    this.defaultPqK = b.defaultPqK;
+    this.defaultGraphDegree = b.defaultGraphDegree;
+    this.defaultOversample = b.defaultOversample;
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/config/WorkerConfig.java:126]: Fixed in `404c1ba`. Added invariant checks for all data-format fallback defaults in `WorkerConfig(Builder)` mirroring `VectorIndexConfig`'s data-parameter validations: `defaultMaxSegmentSize > 0`, `defaultPqM > 0`, `defaultPqK > 1`, `defaultGraphDegree > 0`, `defaultOversample > 0`, `defaultGraphBuildBreadth >= defaultGraphDegree`, `defaultGraphAlpha >= 0`. Added focused tests for each invalid case.

**Comment by panghy** [comment]
All currently raised review items have been addressed on the latest head (`404c1ba`). Both required builds (`ubuntu-latest` and `macos-latest`) completed successfully on that commit. Ready for re-review.

File: src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java:91]: Checked on the current head `cbe8dfc` (and this was already true on `404c1ba`): `VectorIndexConfig` already has an explicit `Objects.requireNonNull(b.defaultThrottle, "defaultThrottle must not be null")` before the `isNegative()` validation. No additional code change was needed for this thread.

[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/config/VectorIndexConfig.java:91]: Checked on the current head `cbe8dfc` (and this was already true on `404c1ba`): `VectorIndexConfig` already has an explicit `Objects.requireNonNull(b.vacuumCooldown, "vacuumCooldown must not be null")` before the `isNegative()` validation. No additional code change was needed for this thread.

File: src/test/java/io/github/panghy/vectorsearch/tasks/GlobalTaskQueueEdgeCaseTest.java [hunk 12]
```diff
@@ -0,0 +1,787 @@
...
+
+  // ---- Test 5: Adapter enqueue wraps correctly ----
+  @Test
+  void buildAdapter_enqueue_wrapsCorrectly() throws Exception {
+    GlobalBuildQueueAdapter adapter = new GlobalBuildQueueAdapter(globalBuildQueue, List.of("my", "index"));
+    BuildTask bt = BuildTask.newBuilder().setSegId(7).build();
+    db.runAsync(tr -> adapter.enqueue(tr, "key2", bt, Duration.ofSeconds(10), Duration.ofMillis(50)))
+        .get(5, TimeUnit.SECONDS);
+    // Allow throttle to elapse before claiming to avoid flakiness under concurrent FDB load
+    Thread.sleep(200);
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/test/java/io/github/panghy/vectorsearch/tasks/GlobalTaskQueueEdgeCaseTest.java]: Fixed in `cbe8dfc`. Removed the `Thread.sleep(200)` from the throttle-related edge-case test and now rely on the existing await/claim path instead of sleeping.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalBuildQueueAdapter.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  @Override
+  public CompletableFuture<TaskKeyMetadata> enqueueIfNotExists(Transaction tx, String key, BuildTask task) {
+    GlobalBuildTask wrapped = GlobalBuildTask.newBuilder()
+        .addAllIndexPath(indexPath)
+        .setTask(task)
+        .build();
+    return globalQueue.enqueueIfNotExists(tx, key, wrapped);
+  }
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalBuildQueueAdapter.java:76]: Fixed in `cbe8dfc`. `GlobalBuildQueueAdapter` now namespaces keys sent to the shared global build queue with the index path, so per-index task idempotency no longer collides across indices.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalMaintenanceQueueAdapter.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  @Override
+  public CompletableFuture<TaskKeyMetadata> enqueueIfNotExists(Transaction tx, String key, MaintenanceTask task) {
+    GlobalMaintenanceTask wrapped = GlobalMaintenanceTask.newBuilder()
+        .addAllIndexPath(indexPath)
+        .setTask(task)
+        .build();
+    return globalQueue.enqueueIfNotExists(tx, key, wrapped);
+  }
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalMaintenanceQueueAdapter.java:72]: Fixed in `cbe8dfc`. `GlobalMaintenanceQueueAdapter` now namespaces keys sent to the shared global maintenance queue with the index path, preventing cross-index collisions on shared maintenance keys.

File: src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+    String key = "compact:" + sorted;
+    MaintenanceTask.Compact c =
+        MaintenanceTask.Compact.newBuilder().addAllSegIds(sorted).build();
+    MaintenanceTask mt = MaintenanceTask.newBuilder().setCompact(c).build();
+    GlobalMaintenanceTask gmt = GlobalMaintenanceTask.newBuilder()
+        .addAllIndexPath(indexPath)
+        .setTask(mt)
+        .build();
+    return db.runAsync(tr -> maintenanceQueue.enqueueIfNotExists(tr, key, gmt))
+        .thenApply(x -> null);
```
[the remark this answers is not in the document]
Comment (panghy) [reply src/main/java/io/github/panghy/vectorsearch/tasks/GlobalWorkerRunner.java:318]: Fixed in `cbe8dfc`. `enqueueCompactOnGlobalQueue(...)` now includes index identity in the compact task key, so compaction tasks from different indices cannot de-duplicate each other on the shared global maintenance queue.

**Comment by panghy** [comment]
Addressed the newly raised review items on the latest head (`cbe8dfc`).

- namespaced global build/maintenance queue keys by index identity
- namespaced compact-task dedupe keys by index identity
- removed the `Thread.sleep(200)` test wait and relied on the existing await/claim path
- re-verified that the `VectorIndexConfig` null checks for `defaultThrottle` and `vacuumCooldown` were already present

Both required builds are now running on `cbe8dfc`. Ready for re-review once CI clears.
